File: src/loom/indexer/adapters/python.py
"""Python language adapter for Loom."""

import logging
from pathlib import Path

import tree_sitter_python as tspy
from tree_sitter import Language, Node, Parser

from loom.store.models import ParsedEdge, Symbol

log = logging.getLogger(__name__)

PY_LANGUAGE = Language(tspy.language())

_PY_EXTENSIONS: frozenset[str] = frozenset({".py", ".pyi"})

_PY_EXCLUDED_DIRS: frozenset[str] = frozenset(
    {".venv", "venv", ".tox", ".mypy_cache", ".pytest_cache"}
)


class PythonAdapter:
    """LanguageAdapter for Python source files."""

    extensions: frozenset[str] = _PY_EXTENSIONS
    language_name: str = "python"
    excluded_dirs: frozenset[str] = _PY_EXCLUDED_DIRS

    def parse(
        self,
        source: bytes,
        file_path: str,
    ) -> tuple[list[Symbol], list[ParsedEdge]]:
        """Parse Python source bytes and extract symbols and edges."""
        suffix = Path(file_path).suffix
        if suffix not in _PY_EXTENSIONS:
            return [], []

        parser = Parser(PY_LANGUAGE)
        tree = parser.parse(source)
        root = tree.root_node

        symbols: list[Symbol] = []
        edges: list[ParsedEdge] = []

        _walk_node(root, source, file_path, symbols, edges, class_stack=[])

        log.debug(
            "Parsed %s: %d symbols, %d edges",
            Path(file_path).name,
            len(symbols),
            len(edges),
        )
        return symbols, edges

    def resolve_module_path(
        self,
        import_path: str,
        source_file: str,
        known_files: set[str],
    ) -> str:
        """Resolve a Python module import path to an actual indexed file.

        Tries direct match, dot-to-slash conversion, and relative imports.
        Returns import_path unchanged if no match found.
        """
        # 1. Direct match
        if import_path in known_files:
            return import_path

        # 2. Relative import — starts with one or more dots
        if import_path.startswith("."):
            dots = len(import_path) - len(import_path.lstrip("."))
            remainder = import_path.lstrip(".")
            # Walk up 'dots' levels from source_file's directory
            base = Path(source_file).parent
            for _ in range(dots - 1):
                base = base.parent
            rel_path = str(base / remainder.replace(".", "/")) if remainder else str(base)
            for candidate in (rel_path + ".py", rel_path + "/__init__.py"):
                if candidate in known_files:
                    return candidate
            return import_path

        # 3. Convert dots to slashes for absolute dotted imports (foo.bar → foo/bar)
        slash_path = import_path.replace(".", "/")
        for candidate in (slash_path + ".py", slash_path + "/__init__.py"):
            if candidate in known_files:
                return candidate

        return import_path


# ── Module-private helpers ───────────────────────────────────────────────────


def _get_text(node: Node, source: bytes) -> str:
    return source[node.start_byte : node.end_byte].decode("utf-8", errors="replace")


def _get_context(source: bytes, node: Node, max_lines: int = 10) -> str:
    start = node.start_point[0]
    end = min(node.end_point[0] + 1, start + max_lines)
    lines = source.split(b"\n")[start:end]
    return b"\n".join(lines).decode("utf-8", errors="replace")


def _child_by_type(node: Node, type_name: str) -> Node | None:
    for child in node.children:
        if child.type == type_name:
            return child
    return None


def _children_by_type(node: Node, type_name: str) -> list[Node]:
    return [c for c in node.children if c.type == type_name]


def _walk_node(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    class_stack: list[str],
) -> None:
    ntype = node.type

    if ntype == "import_statement":
        _handle_import_statement(node, source, file_path, edges)
        return

    if ntype == "import_from_statement":
        _handle_import_from_statement(node, source, file_path, edges)
        return

    if ntype == "decorated_definition":
        _handle_decorated_definition(node, source, file_path, symbols, edges, class_stack)
        return

    if ntype == "function_definition":
        _handle_function_definition(
            node, source, file_path, symbols, edges, class_stack, context=""
        )
        return

    if ntype == "class_definition":
        _handle_class_definition(node, source, file_path, symbols, edges, class_stack, context="")
        return

    if ntype == "expression_statement" and not class_stack:
        _handle_expression_statement(node, source, file_path, symbols)
        return

    for child in node.children:
        _walk_node(child, source, file_path, symbols, edges, class_stack)


def _handle_import_statement(
    node: Node,
    source: bytes,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Handle `import X` and `import X.Y`."""
    for child in node.children:
        if child.type in ("dotted_name", "aliased_import"):
            # For aliased_import, get the first dotted_name child
            if child.type == "aliased_import":
                name_node = _child_by_type(child, "dotted_name")
                if name_node is None:
                    continue
                module_name = _get_text(name_node, source)
            else:
                module_name = _get_text(child, source)
            edges.append(
                ParsedEdge(
                    source_name=module_name,
                    target_name=module_name,
                    target_file=module_name,
                    relationship="imports",
                )
            )


def _handle_import_from_statement(
    node: Node,
    source: bytes,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Handle `from X import Y` and `from . import Y`."""
    module_name: str | None = None
    imported_names: list[str] = []

    for child in node.children:
        if child.type == "dotted_name":
            if module_name is None:
                module_name = _get_text(child, source)
            else:
                imported_names.append(_get_text(child, source))
        elif child.type == "relative_import":
            module_name = _get_text(child, source)
        elif child.type == "import_from_as_names":
            for sub in child.children:
                if sub.type in ("identifier", "dotted_name"):
                    imported_names.append(_get_text(sub, source))
                elif sub.type == "aliased_import":
                    id_node = _child_by_type(sub, "identifier")
                    if id_node:
                        imported_names.append(_get_text(id_node, source))
        elif child.type == "aliased_import":
            id_node = _child_by_type(child, "dotted_name") or _child_by_type(child, "identifier")
            if id_node:
                imported_names.append(_get_text(id_node, source))
        elif child.type == "wildcard_import":
            # `from X import *` — skip, log nothing (acceptable)
            pass

    if module_name is None:
        return

    # Resolve the target file path hint for from-imports
    target_file = module_name

    if imported_names:
        for name in imported_names:
            edges.append(
                ParsedEdge(
                    source_name=name,
                    target_name=name,
                    target_file=target_file,
                    relationship="imports",
                )
            )
    else:
        edges.append(
            ParsedEdge(
                source_name=module_name,
                target_name=module_name,
                target_file=target_file,
                relationship="imports",
            )
        )


def _handle_decorated_definition(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    class_stack: list[str],
) -> None:
    """Handle `@decorator\ndef/class ...`."""
    decorator_text = ""
    for child in node.children:
        if child.type == "decorator":
            # Extract decorator name — first identifier or dotted_name
            for sub in child.children:
                if sub.type in ("identifier", "dotted_name", "call"):
                    decorator_text = _get_text(sub, source)
                    break
        elif child.type == "function_definition":
            _handle_function_definition(
                child, source, file_path, symbols, edges, class_stack, context=decorator_text
            )
        elif child.type == "class_definition":
            _handle_class_definition(
                child, source, file_path, symbols, edges, class_stack, context=decorator_text
            )


def _handle_function_definition(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    class_stack: list[str],
    context: str,
) -> None:
    """Extract a function or method symbol."""
    name_node = _child_by_type(node, "identifier")
    if name_node is None:
        return
    raw_name = _get_text(name_node, source)

    if class_stack:
        full_name = f"{class_stack[-1]}.{raw_name}"
        kind = "method"
    else:
        full_name = raw_name
        kind = "function"

    ctx = context or _get_context(source, node)
    symbols.append(
        Symbol(
            name=full_name,
            kind=kind,
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="python",
            context=ctx,
        )
    )

    # Extract calls from body — do NOT recurse into body for further symbol discovery
    body = _child_by_type(node, "block")
    if body:
        _extract_calls(body, source, full_name, file_path, edges)


def _handle_class_definition(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    class_stack: list[str],
    context: str,
) -> None:
    """Extract a class symbol and recurse into its body."""
    name_node = _child_by_type(node, "identifier")
    if name_node is None:
        return
    class_name = _get_text(name_node, source)

    ctx = context or _get_context(source, node)
    symbols.append(
        Symbol(
            name=class_name,
            kind="class",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="python",
            context=ctx,
        )
    )

    # Extract parent classes from argument_list
    arg_list = _child_by_type(node, "argument_list")
    if arg_list:
        for child in arg_list.children:
            if child.type == "identifier":
                parent_name = _get_text(child, source)
                edges.append(
                    ParsedEdge(
                        source_name=class_name,
                        target_name=parent_name,
                        target_file=None,
                        relationship="extends",
                    )
                )
                edges.append(
                    ParsedEdge(
                        source_name=parent_name,
                        target_name=class_name,
                        target_file=None,
                        relationship="extended_by",
                    )
                )

    # Recurse into body with updated class_stack
    body = _child_by_type(node, "block")
    if body:
        new_stack = class_stack + [class_name]
        for child in body.children:
            _walk_node(child, source, file_path, symbols, edges, new_stack)


def _handle_expression_statement(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
) -> None:
    """Capture module-level UPPER_CASE assignments as variables."""
    for child in node.children:
        if child.type == "assignment":
            lhs = child.children[0] if child.children else None
            if lhs is not None and lhs.type == "identifier":
                name = _get_text(lhs, source)
                if name.isupper() and len(name) >= 1:
                    symbols.append(
                        Symbol(
                            name=name,
                            kind="variable",
                            file=file_path,
                            line=node.start_point[0] + 1,
                            end_line=node.end_point[0] + 1,
                            language="python",
                            context=_get_context(source, node),
                        )
                    )


def _extract_calls(
    node: Node,
    source: bytes,
    caller_name: str,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Recursively extract call edges from a function/method body."""
    if node.type == "call":
        func_node = node.children[0] if node.children else None
        if func_node:
            if func_node.type == "identifier":
                callee = _get_text(func_node, source)
                if callee != caller_name:
                    # If capitalized identifier, likely a class instantiation
                    if callee[0].isupper():
                        edges.append(
                            ParsedEdge(
                                source_name=caller_name,
                                target_name=callee,
                                target_file=None,
                                relationship="instantiates",
                            )
                        )
                    else:
                        edges.append(
                            ParsedEdge(
                                source_name=caller_name,
                                target_name=callee,
                                target_file=None,
                                relationship="calls",
                            )
                        )
            elif func_node.type == "attribute":
                # e.g., self.method(), obj.method()
                attr_children = func_node.children
                if len(attr_children) >= 3:
                    obj_node = attr_children[0]
                    method_node = attr_children[-1]
                    obj_text = _get_text(obj_node, source)
                    method_text = _get_text(method_node, source)
                    callee = f"{obj_text}.{method_text}"
                    if callee != caller_name:
                        edges.append(
                            ParsedEdge(
                                source_name=caller_name,
                                target_name=callee,
                                target_file=None,
                                relationship="calls",
                            )
                        )

    for child in node.children:
        _extract_calls(child, source, caller_name, file_path, edges)

File: crates/loom-core/src/parsers/python.rs
use std::{collections::BTreeSet, path::Path};

use tree_sitter::Node;

use crate::{
    models::ParsedEdge,
    parsers::{
        tree_sitter_utils::{
            child_by_field, child_by_kind, children_by_kind, descendant_of_kind, edge,
            first_named_child_of_kind, language_parse_result, path_parent, push_call_edges,
            push_instantiates, qualified, resolve_direct_or_candidates,
            strip_string_literal_quotes, symbol, text, walk_preorder, Scope,
        },
        LanguageAdapter, ParseResult,
    },
    Result,
};

pub struct PythonAdapter;

impl LanguageAdapter for PythonAdapter {
    fn extensions(&self) -> &'static [&'static str] {
        &[".py", ".pyi"]
    }

    fn language_name(&self) -> &'static str {
        "python"
    }

    fn excluded_dirs(&self) -> &'static [&'static str] {
        &[".venv", "venv", ".tox", ".mypy_cache", ".pytest_cache"]
    }

    fn parse(&self, source: &[u8], file_path: &str) -> Result<ParseResult> {
        let extension = Path::new(file_path)
            .extension()
            .and_then(|extension| extension.to_str())
            .unwrap_or_default();
        if !matches!(extension, "py" | "pyi") {
            return Ok(ParseResult::default());
        }
        language_parse_result(
            source,
            file_path,
            "python",
            tree_sitter_python::LANGUAGE.into(),
            |root, result| walk_node(root, source, file_path, result, Scope::TOP),
        )
    }

    fn resolve_module_path(
        &self,
        import_path: &str,
        source_file: &str,
        known_files: &BTreeSet<String>,
    ) -> String {
        if known_files.contains(import_path) {
            return import_path.to_string();
        }
        if import_path.starts_with('.') {
            let dots = import_path
                .chars()
                .take_while(|character| *character == '.')
                .count();
            let remainder = import_path.trim_start_matches('.').replace('.', "/");
            let mut base = path_parent(source_file);
            for _ in 1..dots {
                base = base.parent().map(Path::to_path_buf).unwrap_or_default();
            }
            let stem = if remainder.is_empty() {
                base.to_string_lossy().into_owned()
            } else {
                base.join(remainder).to_string_lossy().into_owned()
            };
            return resolve_direct_or_candidates(
                import_path,
                known_files,
                [format!("{stem}.py"), format!("{stem}/__init__.py")],
            );
        }
        let slash = import_path.replace('.', "/");
        resolve_direct_or_candidates(
            import_path,
            known_files,
            [format!("{slash}.py"), format!("{slash}/__init__.py")],
        )
    }
}

fn walk_node(
    node: Node<'_>,
    source: &[u8],
    file_path: &str,
    result: &mut ParseResult,
    scope: Scope<'_>,
) {
    match node.kind() {
        "import_statement" => {
            handle_import_statement(node, source, &mut result.edges);
            return;
        }
        "import_from_statement" => {
            handle_import_from_statement(node, source, &mut result.edges);
            return;
        }
        "function_definition" => {
            if let Some(name_node) = child_by_field(node, "name")
                .or_else(|| first_named_child_of_kind(node, &["identifier"]))
            {
                let name = qualified(scope.container, &text(source, name_node));
                let kind = if scope.container.is_some() {
                    "method"
                } else {
                    "function"
                };
                result.symbols.push(symbol(
                    source,
                    node,
                    file_path,
                    "python",
                    name.clone(),
                    kind,
                ));
                push_call_edges(node, source, &name, &mut result.edges, &[]);
                push_instantiates(node, source, &name, &mut result.edges);
            }
            return;
        }
        "class_definition" => {
            if let Some(name_node) = child_by_field(node, "name")
                .or_else(|| first_named_child_of_kind(node, &["identifier"]))
            {
                let name = qualified(scope.container, &text(source, name_node));
                result.symbols.push(symbol(
                    source,
                    node,
                    file_path,
                    "python",
                    name.clone(),
                    "class",
                ));
                handle_bases(node, source, &name, &mut result.edges);
                let mut cursor = node.walk();
                for child in node.children(&mut cursor) {
                    walk_node(
                        child,
                        source,
                        file_path,
                        result,
                        Scope {
                            container: Some(&name),
                            in_function: false,
                        },
                    );
                }
            }
            return;
        }
        "decorated_definition" => {}
        "expression_statement" if scope.container.is_none() && !scope.in_function => {
            handle_upper_assignment(node, source, file_path, result);
        }
        _ => {}
    }
    let next = Scope {
        container: scope.container,
        in_function: scope.in_function || node.kind() == "function_definition",
    };
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        walk_node(child, source, file_path, result, next);
    }
}

fn handle_import_statement(node: Node<'_>, source: &[u8], edges: &mut Vec<ParsedEdge>) {
    walk_preorder(node, &mut |candidate| match candidate.kind() {
        "dotted_name" => {
            let name = text(source, candidate);
            edges.push(edge(name.clone(), name.clone(), "imports", Some(name)));
        }
        "aliased_import" => {
            let module = child_by_kind(candidate, "dotted_name")
                .or_else(|| child_by_kind(candidate, "identifier"))
                .map(|name| text(source, name));
            let alias = children_by_kind(candidate, "identifier")
                .last()
                .map(|name| text(source, name));
            if let Some(module) = module {
                edges.push(edge(
                    alias.unwrap_or_else(|| module.clone()),
                    module.clone(),
                    "imports",
                    Some(module),
                ));
            }
        }
        _ => {}
    });
}

fn handle_import_from_statement(node: Node<'_>, source: &[u8], edges: &mut Vec<ParsedEdge>) {
    if descendant_of_kind(node, &["wildcard_import"]).is_some() {
        return;
    }
    let module = child_by_kind(node, "relative_import")
        .or_else(|| child_by_kind(node, "dotted_name"))
        .map(|module| text(source, module))
        .unwrap_or_default();
    for child in children_by_kind(node, "dotted_name")
        .chain(children_by_kind(node, "identifier"))
        .chain(children_by_kind(node, "aliased_import"))
    {
        let raw = text(source, child);
        if raw == module || raw == "from" || raw == "import" {
            continue;
        }
        let (local, target) = if child.kind() == "aliased_import" {
            let target = child_by_kind(child, "dotted_name")
                .or_else(|| child_by_kind(child, "identifier"))
                .map(|target| text(source, target))
                .unwrap_or_else(|| raw.clone());
            let local = children_by_kind(child, "identifier")
                .last()
                .map(|alias| text(source, alias))
                .unwrap_or_else(|| target.clone());
            (local, target)
        } else {
            (raw.clone(), raw)
        };
        edges.push(edge(local, target, "imports", Some(module.clone())));
    }
}

fn handle_upper_assignment(
    node: Node<'_>,
    source: &[u8],
    file_path: &str,
    result: &mut ParseResult,
) {
    if !matches!(node.kind(), "expression_statement" | "assignment") {
        return;
    }
    let Some(identifier) = descendant_of_kind(node, &["identifier"]) else {
        return;
    };
    let name = text(source, identifier);
    if name
        .chars()
        .all(|character| character.is_uppercase() || character == '_')
    {
        result
            .symbols
            .push(symbol(source, node, file_path, "python", name, "variable"));
    }
}

fn handle_bases(node: Node<'_>, source: &[u8], name: &str, edges: &mut Vec<ParsedEdge>) {
    if let Some(args) = descendant_of_kind(node, &["argument_list"]) {
        for base in children_by_kind(args, "identifier").chain(children_by_kind(args, "attribute"))
        {
            let target = strip_string_literal_quotes(&text(source, base));
            edges.push(edge(name, target.clone(), "extends", None));
            edges.push(edge(target, name, "extended_by", None));
        }
    }
}
